src/QsCompiler/CommandLineTool/CompilationTracker.cs

```diff
@@ -0,0 +1,281 @@
+            var toFindChildrenNodes = new Queue<CompilationTaskNode>();
+
+            // First add the roots (top-level tasks) of all trees to the forest.
+
+            foreach (var entry in CompilationTasks)
+            {
+                if (entry.Value.ParentName == null)
+                {
+                    var node = new CompilationTaskNode(entry.Value);
+                    compilationTasksForest.Add(node);
+                    toFindChildrenNodes.Enqueue(node);
+                }
+            }
+
+            // Iterate through the tasks until all of them have been added to the hierarchy.
+
+            while (toFindChildrenNodes.Count > 0)
+            {
+                var parentNode = toFindChildrenNodes.Dequeue();
+                foreach (var entry in CompilationTasks)
+                {
+                    if (parentNode.Task.Name.Equals(entry.Value.ParentName))
+                    {
+                        var childNode = new CompilationTaskNode(entry.Value);
+                        parentNode.Children.Add(childNode.Task.Name, childNode);
+                        toFindChildrenNodes.Enqueue(childNode);
+                    }
+                }
+            }
+
+            return compilationTasksForest;
+        }
+
+        /// <summary>
+        /// Handles a compilation task start event.
+        /// </summary>
+        private static void CompilationEventStartHandler(CompilationLoader.CompilationTaskEventArgs eventArgs)
+        {
+            string key = CompilationTask.GenerateKey(eventArgs.ParentTaskName, eventArgs.TaskName);
+            if (CompilationTasks.ContainsKey(key))
+            {
+                Warnings.Add(new Warning(WarningType.ProcessAlreadyExists, key));
+                return;
+            }
+
+            CompilationTasks.Add(key, new CompilationTask(eventArgs.ParentTaskName, eventArgs.TaskName));
+        }
+
+        /// <summary>
+        /// Handles a compilation task end event.
+        /// </summary>
+        private static void CompilationEventEndHandler(CompilationLoader.CompilationTaskEventArgs eventArgs)
+        {
+            var key = CompilationTask.GenerateKey(eventArgs.ParentTaskName, eventArgs.TaskName);
+            if (!CompilationTasks.TryGetValue(key, out var task))
+            {
+                Warnings.Add(new Warning(WarningType.ProcessDoesNotExist, key));
+                return;
+            }
+
+            if (!task.IsInProgress())
+            {
+                Warnings.Add(new Warning(WarningType.ProcessAlreadyEnded, key));
+                return;
+            }
+
+            task.End();
+        }
+
+        // Public methods.
+
+        /// <summary>
+        /// Handles a compilation task event.
+        /// </summary>
+        public static void OnCompilationTaskEvent(object sender, CompilationLoader.CompilationTaskEventArgs args)
+        {
+            CompilationEventTypeHandlers[args.Type](args);
```

Comment:
I would go with something like
```suggestion
            if (CompilationEventTypeHandlers.TryGetValue(args.Type, out var handler)) handler(args);
            else Warnings.Add(new Warning(WarningType.UnknownCompilationTaskEventType, args.Type.ToString())); 
```

Reply:
Thanks for the suggestion, this is much better than the original one :)